src/JuliusSweetland.OptiKey/UI/ViewModels/MainViewModel.ServiceEventHandlers.cs

```diff
@@ -2045,5 +1963,85 @@ private void HandleServiceError(object sender, Exception exception)
                 audioService.PlaySound(Settings.Default.ErrorSoundFile, Settings.Default.ErrorSoundVolume);
             }
         }
+
+        private void NavigateToMenu()
+        {
+            Log.Info("Changing keyboard to Menu.");
+            Keyboard = new Menu(CreateBackAction());
+        }
+
+        private void NavigateToVoiceKeyboard()
+        {
+            List<string> voices = GetAvailableVoices();
+
+            if (voices != null && voices.Count > 0)
+            {
+                Log.Info("Changing keyboard to Voice.");
+                Keyboard = new Voice(CreateBackAction(), voices);
+            }
+            else
+            {
+                Log.Warn("No voices available. Returning to menu.");
+                NavigateToMenu();
+            }
+        }
+
+        private void SelectLanguage(Languages language)
+        {
+            Log.Info("Changing keyboard language to " + language);
+            InputService.RequestSuspend(); //Reloading the dictionary locks the UI thread, so suspend input service to prevent accidental selections until complete
+            Settings.Default.KeyboardAndDictionaryLanguage = language;
+            InputService.RequestResume();
+
+            NavigateToVoiceKeyboard();
```

Comment:
Could this be configurable via a new setting so that current behaviour is unchanged. For example the setting could be called "DisplayVoicesWhenChangingKeyboardLanguage" and default it to false (and roaming=true).

Reply:
Absolutely! Would you want a new checkbox in the management console? (Initial thought would be right under "Keyboard Language" on "Words" page.)